app/main.py

```diff
@@ -1,34 +1,128 @@
+        return "Miss!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self,
+                 ships: List[Tuple[Tuple[int, int], Tuple[int, int]]]) -> None:
+        self.field: dict[Tuple[int, int], Ship] = {}
+        self.ships: List[Ship] = []
+        self._validate_field(ships)
+        self._place_ships(ships)
+
+    def _place_ships(
+            self,
+            ships: List[Tuple[Tuple[int, int], Tuple[int, int]]]
+    ) -> None:
+        # Place the ships on the field and store them
+        for ship_coords in ships:
+            start, end = ship_coords
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def _validate_field(
+            self,
+            ships: List[Tuple[Tuple[int, int], Tuple[int, int]]]
+    ) -> None:
+        ship_sizes = {1: 4, 2: 3, 3: 2, 4: 1}
+        ship_counts = {1: 0, 2: 0, 3: 0, 4: 0}
+
+        for ship_coords in ships:
+            start, end = ship_coords
+            if start[0] == end[0]:
+                size = end[1] - start[1] + 1
+            elif start[1] == end[1]:
+                size = end[0] - start[0] + 1
+            else:
+                raise ValueError("Ships must be either "
+                                 "horizontal or vertical.")
+
+            if size not in ship_counts:
+                raise ValueError("Invalid ship size.")
+            ship_counts[size] += 1
+
+        # Check if the ship sizes match the expected counts
+        for size, count in ship_counts.items():
+            if count != ship_sizes[size]:
+                raise ValueError(f"Invalid number of ships with {size} decks.")
+
+        # Check that ships are not in neighboring cells
+        for ship_coords in ships:
+            for (row, col) in self._get_adjacent_cells(ship_coords):
+                if (row, col) in self.field:
+                    raise ValueError("Ships cannot be in neighboring cells.")
+
+    def _get_adjacent_cells(
+            self,
+            ship_coords: Tuple[Tuple[int, int], Tuple[int, int]]
+    ) -> set:
+        start, end = ship_coords
+        adjacent_cells = set()
+
+        # Determine all adjacent cells around the ship
+        for row in range(start[0] - 1, end[0] + 2):
+            for col in range(start[1] - 1, end[1] + 2):
+                if ((row, col) != (start[0], start[1]) and (row, col)
+                        != (end[0], end[1])):
```

Comment:
The condition here is incorrect for determining adjacent cells. It should include the start and end cells as well, otherwise, it will exclude the actual ship cells from being considered adjacent. Consider revising the logic to correctly identify all adjacent cells.